File: scripts/run_python_package_tests.mjs
#!/usr/bin/env node

import { spawnSync } from "node:child_process";
import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";

const root = resolve(dirname(fileURLToPath(import.meta.url)), "..");
const candidates = process.platform === "win32"
  ? [
      ["py", ["-3"]],
      ["python", []],
      ["python3", []],
    ]
  : [
      ["python3", []],
      ["python", []],
    ];

function run(command, args, options = {}) {
  return spawnSync(command, args, {
    cwd: root,
    stdio: "inherit",
    shell: false,
    ...options,
  });
}

function findPython() {
  const versionCheck = "import sys; raise SystemExit(0 if sys.version_info >= (3, 10) else 1)";
  for (const [command, prefixArgs] of candidates) {
    const args = [...prefixArgs, "-c", versionCheck];
    const result = spawnSync(command, args, {
      cwd: root,
      stdio: "ignore",
      shell: false,
    });
    if (result.status === 0) {
      return [command, prefixArgs];
    }
  }
  throw new Error("Python 3.10 or newer executable not found.");
}

const [pythonCommand, pythonArgs] = findPython();
const result = run(
  pythonCommand,
  [...pythonArgs, "scripts/run_python_package_tests.py"],
);

process.exit(result.status ?? 1);

File: scripts/run_python_package_tests.py
#!/usr/bin/env python3
"""Run Python helper package tests with a local source import path."""

from __future__ import annotations

import os
from pathlib import Path
import subprocess
import sys


ROOT = Path(__file__).resolve().parents[1]
PYTHON_SRC = ROOT / "packages/python/src"


def run(command: list[str], env: dict[str, str]) -> None:
    subprocess.run(command, cwd=ROOT, env=env, check=True)


def main() -> int:
    env = os.environ.copy()
    existing_pythonpath = env.get("PYTHONPATH")
    env["PYTHONPATH"] = (
        str(PYTHON_SRC)
        if not existing_pythonpath
        else str(PYTHON_SRC) + os.pathsep + existing_pythonpath
    )
    run(
        [
            sys.executable,
            "-m",
            "unittest",
            "discover",
            "-s",
            str(ROOT / "packages/python/tests"),
        ],
        env,
    )
    run([sys.executable, str(ROOT / "scripts/check_python_package_artifacts.py")], env)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
